File: project/two_pianos/lib/two_pianos_web/channels/room_channel.ex
defmodule TwoPianosWeb.RoomChannel do

  use TwoPianosWeb, :channel

  def join("room:" <> room_id, _, socket) do

    room_id_atom = room_id |> String.to_atom()

    if RoomManager.room_exists?(room_id_atom) do
      room_id_atom |> attempt_join(socket)
    else
      {:error, %{reason: "Room " <> room_id <> " doesn't exist"}}
    end

  end

  def handle_in("keys_pressed", %{keys_pressed: keys}, socket) do
    socket |> broadcast!("press", keys)
  end

  def handle_in("keys_released", %{keys_released: keys}, socket) do
    socket |> broadcast!("release", keys)
  end

  def handle_in("message_posted", %{message_posted: message}, socket) do
    socket |> broadcast!("chat_message", message)
  end

  # Helpers

  def attempt_join(room_id, socket) do

    case Room.join(room_id, socket.assigns[:user_id]) do
      :success           -> {:ok,    socket}
      {:failure, reason} -> {:error, %{reason: reason}}
    end

  end

end

File: project/two_pianos/assets/js/room_channel.js
import { Piano } from "./piano"

export class RoomChannel{

    static setup_and_join(socket, user_id, room_id, their_piano){

        let room_channel = socket.channel(`room:${room_id}`)
        let keycodes     = Piano.get_keycodes()

        document.addEventListener('onkeydown', function(e){
            e = e || window.event
    
            if (!e.repeat && e.keyCode in keycodes && !(input === document.activeElement)){
                room_channel.push("keys_pressed", {keys_pressed: keycodes[e.keyCode]})
            }
        })

        document.addEventListener('onkeyup', function(e){
            e = e || window.event
    
            if (e.keyCode in keycodes && !(input === document.activeElement)){
                room_channel.push("keys_released", {keys_released: keycodes[e.keyCode]})
            }
        })

        document.getElementById("message-post-btn").addEventListener("click", e => {

            let message = document.getElementById("message-input").value

            if (message !== ""){
                room_channel.push("message_posted", {message_posted: message})
                RoomChannel.render_message("you", message)
            }

        })

        room_channel.on("keys_pressed", (resp) => {
            their_piano.press(resp.keys_pressed)
        })
            
        room_channel.on("keys_released", (resp) => {
            their_piano.release(resp.keys_released)
        })

        room_channel.on("message_posted", (resp) => {
            RoomChannel.render_message("them", resp.message_posted)
        })

        room_channel.join()
            .receive("ok", resp => {
                console.log("Room Channel: Joined successfully", resp) 
            })
            .receive("error", resp => { 
                console.log("Room Channel: Unable to join", resp) 
            })
            
        return room_channel
        
    }

    static render_message(you_or_them, message) {
        let chat_box       = document.getElementById("chat-box")
        let message_block  = document.createElement('p')

        message_block.insertAdjacentHTML('afterbegin', `<b>${you_or_them === 'you' ? '(:' : ':)'}</b>${message}`)
        chat_box.appendChild(message_block)
    }

}